ksp_plugin_adapter/loader.cs

```
@@ -81,33 +64,44 @@ internal static string LoadPrincipiaDllAndInitGoogleLogging() {
       }
     }
     try {
-      if (Environment.OSVersion.Platform == PlatformID.Win32NT) {
-        principia_dll_ = LoadLibrary(possible_dll_paths[0]);
-      } else {
-        const int RTLD_NOW = 2;
-        const int RTLD_GLOBAL = 8;
-        foreach (string path in possible_dll_paths) {
-          principia_dll_ = dlopen(path, RTLD_NOW|RTLD_GLOBAL);
-          if (principia_dll_ != IntPtr.Zero) {
-            break;
-          }
+      // First try to load the x64 DLL.  This should always work, and it makes
+      // it possible to query the CPU flags.  Note that we can only call a
+      // limited set of function before we load the final DLL.  In particular,
+      // we cannot `InitGoogleLogging` as this seems to reference dangling
+      // pointers on the second load.
+      string error_x64 = LoadPrincipiaDllForPlatform(
+          possible_dll_path_patterns,
+          dll_filename: dll_filename,
+          platform: "x64");
+      if (error_x64 != null) {
+        return error_x64;
+      }
+      Interface.GetCPUIDFeatureFlags(out bool has_avx, out bool has_fma);
+      if (has_fma && !has_avx) {
+        return "Principia does not run on processors with FMA support but no " +
+               "AVX support.";
+      } else if (has_fma) {
+        // If it turns out that the machine has FMA (and therefore AVX), change
+        // our mind and load the x64_AVX_FMA DLL.  It's necessary to unload the
+        // DLL, otherwise the base address reported by `InitGoogleLogging` might
+        // (incorrectly) be the one of the `x64` DLL.
+        UnloadPrincipiaDll(platform: "x64");
```

Why are we discarding the error message from UnloadPrincipiaDll?